fedimint-core/src/task/jit.rs

```diff
@@ -0,0 +1,188 @@
+        }
+    }
+}
+
+/// A value that initializes eagerly in parallel in a falliable way
+#[derive(Debug, Clone)]
+pub struct JitCore<T, E> {
+    // on last drop it lets the inner task to know it should stop, because we don't care anymore
+    _cancel_tx: sync::mpsc::Sender<()>,
+    // since joinhandles, are not portable, we use this to wait for the value
+    // Note: the `std::sync::Mutex` is used intentionally, since contention never actually happens
+    // and we want to avoid cancelation because of it
+    #[allow(clippy::type_complexity)]
+    val_rx:
+        Arc<std::sync::Mutex<Option<tokio::sync::oneshot::Receiver<std::result::Result<T, E>>>>>,
+    val: sync::OnceCell<T>,
+}
+
+impl<T, E> JitCore<T, E>
+where
+    T: MaybeSend + 'static,
+    E: MaybeSend + 'static,
+{
+    /// Create `JitTry` value, and spawn a future `f` that computes its value
+    ///
+    /// Unlike normal Rust futures, the `f` executes eagerly (is spawned as a
+    /// tokio task).
+    pub fn new_try<Fut>(f: impl FnOnce() -> Fut + 'static + MaybeSend) -> Self
+    where
+        Fut: Future<Output = std::result::Result<T, E>> + 'static + MaybeSend,
+    {
+        let (cancel_tx, mut cancel_rx) = tokio::sync::mpsc::channel(1);
+        let (val_tx, val_rx) = oneshot::channel();
+        super::imp::spawn(
+            &format!("JitTry {} value", std::any::type_name::<T>()),
+            async move {
+                select! {
+                    _ = cancel_rx.recv() => {
+                        trace!(target: LOG_TASK, r#type = %std::any::type_name::<T>(), "JitTry value future canceled");
+                    },
+                    val = f() => {
+                        match val_tx.send(val) {
+                            Ok(_) => {
+                                trace!(target: LOG_TASK, r#type = %std::any::type_name::<T>(), "JitTry value ready");
+                            },
+                            Err(_) => {
+                                trace!(target: LOG_TASK,  r#type = %std::any::type_name::<T>(), "JitTry value ready, but ignored");
+                            },
+                        };
+                    },
+                }
+            },
+        )
+        .expect("spawn not fail");
+
+        Self {
+            _cancel_tx: cancel_tx,
+            val_rx: Arc::new(Some(val_rx).into()),
+            val: sync::OnceCell::new(),
+        }
+    }
+
+    /// Get the reference to the value, potentially blocking for the
+    /// initialization future to complete
+    pub async fn get_try(&self) -> std::result::Result<&T, OneTimeError<E>> {
+        #[allow(clippy::await_holding_lock)]
+        self.val
+            .get_or_try_init(|| async {
+                let val_res = self
+                    .val_rx
+                    // this lock gets locked only once so it's kind of useless other than making
+                    // Rust happy, but the overhead doesn't matter
+                    .lock()
```

Comment:
We could use `try_lock` here to enforce the invariant described above. This will avoid any accidental tokio executor lockups by panicking instead.